pkg/query/aggregation_label_rewriter.go

```diff
@@ -0,0 +1,137 @@
+	aggregationLabelRewriterRuntimeSeconds *prometheus.HistogramVec
+}
+
+func newAggregationLabelRewriterMetrics(reg prometheus.Registerer, desiredLabelValue string) *aggregationLabelRewriterMetrics {
+	m := &aggregationLabelRewriterMetrics{}
+
+	m.aggregationLabelRewriteSkippedCount = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "skipped_total",
+		Help: "Total number of times aggregation-label-rewriter was enabled but skipped",
+	}, []string{"reason"})
+	m.aggregationLabelRewrittenCount = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name:        "label_rewritten_total",
+		Help:        "Total number of times aggregation-label-rewriter rewrote the aggregation label",
+		ConstLabels: prometheus.Labels{"new_value": desiredLabelValue},
+	}, []string{"old_value"})
+	m.aggregationLabelAddedCount = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "label_added_total",
+		Help:        "Total number of times aggregation-label-rewriter added the aggregation label",
+		ConstLabels: prometheus.Labels{"new_value": desiredLabelValue},
+	})
+	m.aggregationLabelRewriterRuntimeSeconds = promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+		Name: "total_runtime_seconds",
+		Help: "Runtime of aggregation-label-rewriter in seconds",
+	}, []string{"is_rewritten"})
+
+	return m
+}
+
+func NewNopAggregationLabelRewriter() *AggregationLabelRewriter {
+	return &AggregationLabelRewriter{
+		enabled: false,
+	}
+}
+
+func NewAggregationLabelRewriter(logger log.Logger, reg prometheus.Registerer, desiredLabelValue string) *AggregationLabelRewriter {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	g := glob.MustCompile(aggregatedMetricNameGlobPattern)
+	return &AggregationLabelRewriter{
+		enabled:                  desiredLabelValue != "",
+		logger:                   logger,
+		metrics:                  newAggregationLabelRewriterMetrics(reg, desiredLabelValue),
+		desiredLabelValue:        desiredLabelValue,
+		aggregatedMetricNameGlob: g,
+	}
+}
+
+func (a *AggregationLabelRewriter) Rewrite(ms []*labels.Matcher) []*labels.Matcher {
+	if a.enabled {
+		startOfRun := time.Now()
+		needsRewrite := false
+		skipReason := "no-name-matcher"
+		var aggregationLabelMatcher *labels.Matcher
+		aggregationLabelIndex := -1
+		for i := 0; i < len(ms); i++ {
+			m := ms[i]
+			if !needsRewrite && m.Name == labels.MetricName {
+				if m.Type == labels.MatchEqual {
+					if a.aggregatedMetricNameGlob.Match(m.Value) {
+						needsRewrite = true
+					} else {
+						skipReason = "not-aggregated-metric"
+					}
+				} else {
+					skipReason = "not-using-equal-name-match"
+					level.Debug(a.logger).Log("msg", "skipped due to not using equal in name matcher", "matcher", m)
+					break
+				}
+			} else if needsRewrite && aggregationLabelMatcher != nil {
+				break
+			} else if m.Name == aggregationLabelName {
+				aggregationLabelMatcher = m
+				aggregationLabelIndex = i
+			}
```

Comment:
nit: for readability, consider short circuit first instead of nested if-else branches